File: apple_fm_audit/store.py
"""SQLite log of proxied HTTP calls."""

from __future__ import annotations

import json
import sqlite3
import threading
import time
from typing import Any

from apple_fm_audit.annotate import annotate_row


def _decode(body: bytes) -> str:
    if not body:
        return ""
    return body.decode("utf-8", errors="replace")


class Store:
    def __init__(self, path: str):
        self.path = path
        self._lock = threading.Lock()
        self.conn = sqlite3.connect(path, check_same_thread=False)
        self.conn.row_factory = sqlite3.Row
        self.conn.execute(
            """
            CREATE TABLE IF NOT EXISTS calls (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                ts REAL NOT NULL,
                method TEXT NOT NULL,
                path TEXT NOT NULL,
                query TEXT NOT NULL DEFAULT '',
                status INTEGER,
                duration_ms INTEGER,
                req_headers TEXT NOT NULL DEFAULT '{}',
                req_body TEXT NOT NULL DEFAULT '',
                res_headers TEXT NOT NULL DEFAULT '{}',
                res_body TEXT NOT NULL DEFAULT '',
                error TEXT
            )
            """
        )
        self.conn.execute(
            "CREATE INDEX IF NOT EXISTS calls_ts ON calls(ts DESC)"
        )
        self.conn.commit()

    def close(self) -> None:
        self.conn.close()

    def insert(
        self,
        *,
        method: str,
        path: str,
        query: str,
        status: int | None,
        duration_ms: int | None,
        req_headers: dict[str, str],
        req_body: bytes,
        res_headers: dict[str, str],
        res_body: bytes,
        error: str | None,
    ) -> int:
        with self._lock:
            cur = self.conn.execute(
            """
            INSERT INTO calls (
                ts, method, path, query, status, duration_ms,
                req_headers, req_body, res_headers, res_body, error
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                time.time(),
                method,
                path,
                query or "",
                status,
                duration_ms,
                json.dumps(req_headers),
                _decode(req_body),
                json.dumps(res_headers),
                _decode(res_body),
                error,
            ),
        )
            self.conn.commit()
            return int(cur.lastrowid)

    def list_calls(
        self, path_contains: str | None = None, limit: int = 200
    ) -> list[dict[str, Any]]:
        sql = (
            "SELECT id, ts, method, path, query, status, duration_ms, error, res_body "
            "FROM calls"
        )
        args: list[Any] = []
        if path_contains:
            sql += " WHERE path LIKE ?"
            args.append(f"%{path_contains}%")
        sql += " ORDER BY id DESC LIMIT ?"
        args.append(limit)
        with self._lock:
            rows = self.conn.execute(sql, args).fetchall()
        return [annotate_row(dict(r)) for r in rows]

    def get_call(self, call_id: int) -> dict[str, Any] | None:
        with self._lock:
            row = self.conn.execute(
                "SELECT * FROM calls WHERE id = ?", (call_id,)
            ).fetchone()
        if row is None:
            return None
        data = dict(row)
        extra = annotate_row(
            {
                "status": data.get("status"),
                "error": data.get("error"),
                "res_body": data.get("res_body"),
            }
        )
        data["prompt_tokens"] = extra["prompt_tokens"]
        data["completion_tokens"] = extra["completion_tokens"]
        data["total_tokens"] = extra["total_tokens"]
        data["issue"] = extra["issue"]
        data["issue_label"] = extra["issue_label"]
        return data

    def clear(self) -> None:
        with self._lock:
            self.conn.execute("DELETE FROM calls")
            self.conn.commit()

File: Sources/AppleFMAuditCore/Store.swift
import Foundation
import SQLite3

final class DBHandle: @unchecked Sendable {
    var pointer: OpaquePointer?
    init(_ pointer: OpaquePointer?) {
        self.pointer = pointer
    }
    deinit {
        if let ptr = pointer {
            sqlite3_close(ptr)
        }
    }
}

public actor Store {
    private let dbHandle: DBHandle
    public let path: String

    private var db: OpaquePointer? {
        dbHandle.pointer
    }

    public init(path: String) throws {
        self.path = path
        if path != ":memory:" {
            let dir = URL(fileURLWithPath: path).deletingLastPathComponent()
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }

        var handle: OpaquePointer?
        let flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX
        let rc = sqlite3_open_v2(path, &handle, flags, nil)
        guard rc == SQLITE_OK, let dbPtr = handle else {
            let msg = handle != nil ? String(cString: sqlite3_errmsg(handle)) : "Failed to open sqlite"
            sqlite3_close(handle)
            throw NSError(domain: "Store", code: Int(rc), userInfo: [NSLocalizedDescriptionKey: msg])
        }
        self.dbHandle = DBHandle(dbPtr)

        try Self.execute(db: dbPtr, sql: """
            CREATE TABLE IF NOT EXISTS calls (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                ts REAL NOT NULL,
                method TEXT NOT NULL,
                path TEXT NOT NULL,
                query TEXT NOT NULL DEFAULT '',
                status INTEGER,
                duration_ms INTEGER,
                req_headers TEXT NOT NULL DEFAULT '{}',
                req_body TEXT NOT NULL DEFAULT '',
                res_headers TEXT NOT NULL DEFAULT '{}',
                res_body TEXT NOT NULL DEFAULT '',
                error TEXT
            );
            CREATE INDEX IF NOT EXISTS calls_ts ON calls(ts DESC);
        """)
    }

    public func close() {
        if let ptr = dbHandle.pointer {
            sqlite3_close(ptr)
            dbHandle.pointer = nil
        }
    }

    private static func execute(db: OpaquePointer?, sql: String) throws {
        guard let db = db else { throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: "Database closed"]) }
        var err: UnsafeMutablePointer<CChar>?
        if sqlite3_exec(db, sql, nil, nil, &err) != SQLITE_OK {
            let msg = err != nil ? String(cString: err!) : "Unknown exec error"
            sqlite3_free(err)
            throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: msg])
        }
    }

    private func execute(sql: String) throws {
        try Self.execute(db: self.db, sql: sql)
    }

    @discardableResult
    public func insert(
        method: String,
        path: String,
        query: String = "",
        status: Int?,
        durationMs: Int?,
        reqHeaders: String,
        reqBody: String,
        resHeaders: String,
        resBody: String,
        error: String?
    ) throws -> Int64 {
        guard let db = db else { throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: "Database closed"]) }
        let sql = """
            INSERT INTO calls (ts, method, path, query, status, duration_ms, req_headers, req_body, res_headers, res_body, error)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: String(cString: sqlite3_errmsg(db))])
        }
        defer { sqlite3_finalize(stmt) }

        let now = Date().timeIntervalSince1970
        sqlite3_bind_double(stmt, 1, now)
        sqlite3_bind_text(stmt, 2, (method as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 3, (path as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 4, (query as NSString).utf8String, -1, nil)

        if let status = status {
            sqlite3_bind_int(stmt, 5, Int32(status))
        } else {
            sqlite3_bind_null(stmt, 5)
        }

        if let durationMs = durationMs {
            sqlite3_bind_int(stmt, 6, Int32(durationMs))
        } else {
            sqlite3_bind_null(stmt, 6)
        }

        sqlite3_bind_text(stmt, 7, (reqHeaders as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 8, (reqBody as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 9, (resHeaders as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 10, (resBody as NSString).utf8String, -1, nil)

        if let error = error {
            sqlite3_bind_text(stmt, 11, (error as NSString).utf8String, -1, nil)
        } else {
            sqlite3_bind_null(stmt, 11)
        }

        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: String(cString: sqlite3_errmsg(db))])
        }

        return sqlite3_last_insert_rowid(db)
    }

    public func listCalls(pathContains: String? = nil, limit: Int = 200) throws -> [CallSummary] {
        guard let db = db else { throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: "Database closed"]) }
        var sql = "SELECT id, ts, method, path, query, status, duration_ms, error, res_body FROM calls"
        if pathContains != nil {
            sql += " WHERE path LIKE ?"
        }
        sql += " ORDER BY id DESC LIMIT ?"

        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: String(cString: sqlite3_errmsg(db))])
        }
        defer { sqlite3_finalize(stmt) }

        var bindIdx: Int32 = 1
        if let pattern = pathContains {
            let needle = "%\(pattern)%"
            sqlite3_bind_text(stmt, bindIdx, (needle as NSString).utf8String, -1, nil)
            bindIdx += 1
        }
        sqlite3_bind_int(stmt, bindIdx, Int32(limit))

        var results: [CallSummary] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            let id = sqlite3_column_int64(stmt, 0)
            let ts = sqlite3_column_double(stmt, 1)
            let method = String(cString: sqlite3_column_text(stmt, 2))
            let path = String(cString: sqlite3_column_text(stmt, 3))
            let query = sqlite3_column_text(stmt, 4).map { String(cString: $0) } ?? ""
            let status: Int? = sqlite3_column_type(stmt, 5) == SQLITE_NULL ? nil : Int(sqlite3_column_int(stmt, 5))
            let durationMs: Int? = sqlite3_column_type(stmt, 6) == SQLITE_NULL ? nil : Int(sqlite3_column_int(stmt, 6))
            let error: String? = sqlite3_column_text(stmt, 7).map { String(cString: $0) }
            let resBody: String? = sqlite3_column_text(stmt, 8).map { String(cString: $0) }

            let usage = Annotate.extractUsage(resBody: resBody)
            let issue = Annotate.classifyIssue(status: status, resBody: resBody, error: error)

            results.append(CallSummary(
                id: id,
                ts: ts,
                method: method,
                path: path,
                query: query,
                status: status,
                durationMs: durationMs,
                error: error,
                promptTokens: usage.prompt,
                completionTokens: usage.completion,
                totalTokens: usage.total,
                issue: issue.id,
                issueLabel: issue.label
            ))
        }

        return results
    }

    public func getCall(id: Int64) throws -> CallDetail? {
        guard let db = db else { throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: "Database closed"]) }
        let sql = "SELECT id, ts, method, path, query, status, duration_ms, req_headers, req_body, res_headers, res_body, error FROM calls WHERE id = ? LIMIT 1;"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw NSError(domain: "Store", code: -1, userInfo: [NSLocalizedDescriptionKey: String(cString: sqlite3_errmsg(db))])
        }
        defer { sqlite3_finalize(stmt) }

        sqlite3_bind_int64(stmt, 1, id)

        guard sqlite3_step(stmt) == SQLITE_ROW else {
            return nil
        }

        let cid = sqlite3_column_int64(stmt, 0)
        let ts = sqlite3_column_double(stmt, 1)
        let method = String(cString: sqlite3_column_text(stmt, 2))
        let path = String(cString: sqlite3_column_text(stmt, 3))
        let query = sqlite3_column_text(stmt, 4).map { String(cString: $0) } ?? ""
        let status: Int? = sqlite3_column_type(stmt, 5) == SQLITE_NULL ? nil : Int(sqlite3_column_int(stmt, 5))
        let durationMs: Int? = sqlite3_column_type(stmt, 6) == SQLITE_NULL ? nil : Int(sqlite3_column_int(stmt, 6))
        let reqHeaders = sqlite3_column_text(stmt, 7).map { String(cString: $0) } ?? "{}"
        let reqBody = sqlite3_column_text(stmt, 8).map { String(cString: $0) } ?? ""
        let resHeaders = sqlite3_column_text(stmt, 9).map { String(cString: $0) } ?? "{}"
        let resBody = sqlite3_column_text(stmt, 10).map { String(cString: $0) } ?? ""
        let error = sqlite3_column_text(stmt, 11).map { String(cString: $0) }

        let usage = Annotate.extractUsage(resBody: resBody)
        let issue = Annotate.classifyIssue(status: status, resBody: resBody, error: error)

        return CallDetail(
            id: cid,
            ts: ts,
            method: method,
            path: path,
            query: query,
            status: status,
            durationMs: durationMs,
            reqHeaders: reqHeaders,
            reqBody: reqBody,
            resHeaders: resHeaders,
            resBody: resBody,
            error: error,
            promptTokens: usage.prompt,
            completionTokens: usage.completion,
            totalTokens: usage.total,
            issue: issue.id,
            issueLabel: issue.label
        )
    }

    public func clear() throws {
        try execute(sql: "DELETE FROM calls;")
    }
}
